pennylane/wires.py

```diff
@@ -0,0 +1,322 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`Wires` class, which takes care of wire bookkeeping.
+"""
+from collections import Sequence, Iterable
+import numpy as np  # for random functions
+from copy import deepcopy
+
+# tolerance for integer-like wire indices
+TOLERANCE = 1e-8
+
+
+class WireError(Exception):
+    """Exception raised by a :class:`~.pennylane.wires.Wire` object when it is unable to process wires.
+    """
+
+
+def _clean(iterable):
+    """Converts 'iterable' into list of integers, and checks that all integers are unique and non-negative.
```

Comment:
Does this allow for nested lists of wires? E.g.,
```python
>>> ob = qml.PauliX(0) @ qml.Hermitian(A, [1, 4]) @ qml.PauliZ(2)
[[0], [1, 4], [2]]
```

Reply:
No, good point! Can't one have a list of wire objects? I will need to check!

Reply:
This will help solve a major pain point with the `Tensor` class, as logic always has to use if statements to deal with a sudden nested list. Or if the nesting doesn't matter, it has to flatten the wires list.

Maybe the wires class can deal with this in a smarter manner.